lib/srv/db/mongodb/autousers_admin.go

```diff
@@ -0,0 +1,260 @@
+	return shareableClient, trace.Wrap(err)
+}
+
+func getShareableAdminClient(ctx context.Context, cache *utils.FnCache, sessionCtx *common.Session, e *Engine, makeBasicClient makeBasicAdminClientFunc) (*shareableAdminClient, error) {
+	key := struct {
+		databaseName string
+		databaseURI  string
+		adminUser    string
+	}{
+		databaseName: sessionCtx.Database.GetName(),
+		adminUser:    sessionCtx.Database.GetAdminUser().Name,
+		// Just in case the same database name is updated to a different database server.
+		databaseURI: sessionCtx.Database.GetURI(),
+	}
+
+	shareableClient, err := utils.FnCacheGet(ctx, cache, key, func(ctx context.Context) (*shareableAdminClient, error) {
+		rawClient, err := makeBasicClient(ctx, sessionCtx, e)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		shareableClient, err := newShareableAdminClient(shareableAdminClientConfig{
+			rawClient:    rawClient,
+			adminUser:    sessionCtx.Database.GetAdminUser().Name,
+			databaseName: sessionCtx.Database.GetName(),
+			log:          e.Log,
+			clock:        e.Clock,
+		})
+		return shareableClient, trace.Wrap(err)
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return shareableClient.acquire(), nil
+}
+
+type shareableAdminClientConfig struct {
+	rawClient    adminClient
+	adminUser    string
+	databaseName string
+	clock        clockwork.Clock
+	log          logrus.FieldLogger
+	cleanupTTL   time.Duration
+}
+
+// checkAndSetDefaults validates config and sets defaults.
+func (c *shareableAdminClientConfig) checkAndSetDefaults() error {
+	if c.rawClient == nil {
+		return trace.BadParameter("missing mongo.Client")
+	}
+	if c.adminUser == "" {
+		return trace.BadParameter("missing adminUser")
+	}
+	if c.databaseName == "" {
+		return trace.BadParameter("missing databaseName")
+	}
+	if c.clock == nil {
+		c.clock = clockwork.NewRealClock()
+	}
+	if c.log == nil {
+		c.log = logrus.StandardLogger()
+	}
+	if c.cleanupTTL <= 0 {
+		c.cleanupTTL = adminClientCleanupTTL
+	}
+	return nil
+}
+
+type shareableAdminClient struct {
+	adminClient
+	shareableAdminClientConfig
+	wg    sync.WaitGroup
+	timer clockwork.Timer
+}
+
+func newShareableAdminClient(cfg shareableAdminClientConfig) (*shareableAdminClient, error) {
+	if err := cfg.checkAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	c := &shareableAdminClient{
+		adminClient:                cfg.rawClient,
+		shareableAdminClientConfig: cfg,
+		timer:                      cfg.clock.NewTimer(cfg.cleanupTTL),
+	}
+
+	go c.waitAndCleanup()
+	return c, nil
+}
+
+func (c *shareableAdminClient) waitAndCleanup() {
+	c.log.Debugf("Created new MongoDB connection as admin user %q on %q.", c.adminUser, c.databaseName)
+
+	// Wait until TTL.
+	<-c.timer.Chan()
+
+	// Wait until all shared sessions are released.
+	c.wg.Wait()
+
+	// Disconnect.
+	if err := c.adminClient.Disconnect(context.Background()); err != nil {
```

Comment:
nit: Can we update the comment for a reader: 
```suggestion
	// Disconnect connection. This happens after cache item expired to ensure that shared client won't be reused when wrapped client was disconnected.  
	if err := c.adminClient.Disconnect(context.Background()); err != nil {
```